make OMP_NUM_THREADS default in launch.py

zhaojuanmao opened the pull request:
Summary:
per https://github.com/pytorch/pytorch/issues/22260, default number of open mp threads are spawned to be the same of number of cores available, for multi processing data parallel cases, too many threads may be spawned and could overload the CPU, resulting in performance regression.

so set OMP_NUM_THREADS = number of CPU processors/number of processes in default to neither overload or waste CPU threads

Differential Revision: D16092225

soumith reviewed (changes requested):
the script `torch/distributed/examples/launch_script_example.py` has no place in pytorch core.
If you want to, add it in https://github.com/pytorch/examples/ separately

xuhdev commented on torch/distributed/launch.py:
```diff
@@ -203,6 +207,9 @@ def main():
     current_env["WORLD_SIZE"] = str(dist_world_size)
 
+    if 'OMP_NUM_THREADS' not in os.environ:
+        current_env["OMP_NUM_THREADS"] = str(int(multiprocessing.cpu_count() / args.nproc_per_node))
```
There would still be waste if `multiprocessing.cpu_count() / args.nproc_per_node` is not an integer. I would move the setting of `OMP_NUM_THREAD` down into the loop, and have the number of threads be `multiprocessing.cpu_count() / args.nproc_per_node + 1` if `local_rank < multiprocessing.cpu_count() % args.nproc_per_node`.

fmassa replied:
The optimal number of OMP threads might not simply be nb of CPU / nb of processes per node, because there is also data loader that should be taken into account.

See https://github.com/pytorch/vision/pull/818#issuecomment-508465650 for example.

This is something that we should raise awareness of in some way, or else people will have much slower training times for some models and they won't know why.

pietern replied:
Very good point. We could choose to log the default value, just to call attention that the value is set in the first place. Then, if people have performance issues, there is at least a hint about where to look.

mfuntowicz commented:
Pushing forward the discussion regarding launch script, is OMP_NUM_THREADS the only variable that needs to be taken into account ?

Especially, I'm thinking about NUMA setups, we're GPU/processes should be dispatched w.r.t CPU affinity.

One thing might be to add NUMA detection inside the script and then set OMP_PROC_BIND=true and possibly some KMP_AFFINITY ... What do you think ?

fmassa commented:
Thumbs up for merging such a feature rather quickly.

I've just stumbled upon https://github.com/pytorch/vision/pull/818#issuecomment-508465650, where not setting `OMP_NUM_THREADS` can lead to up to 3x slowdown in some trainings.

pietern commented on torch/distributed/launch.py:
```diff
@@ -1,3 +1,6 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+
```
These imports are not necessary.

zhaojuanmao replied:
looks like there were lint errors in both fbcode and OSS, so I'm adding it back

zhaojuanmao commented:
Thanks for all your comments, as suggested, I will
1) print the OMP_NUM_THREADS value to call users' attention, so that users can further tune it based their application as needed(e.g, @fmassa mentioned data loader impact the optimal value in his application) 
2) I will make changes as what @xuhdev's suggested
3) this diff will focus only on OMP_NUM_THREADS variable, tuning other variables need more thoughts and discussions

pietern reviewed (commented):
Looks good to me. 

FWIW, @zhaojuanmao I think that having different values for different processes will be harmful, in that there is potential for a performance imbalance. I expect 99% of the runs launched through this tool to use distributed data parallel, where the faster processes have to wait on the slower ones anyway. I would recommend removing the logic that gives some ranks an extra thread. Then, in order to make this work on machines with fewer CPUs than launched processes, there should also be a `max(1, avg_threads_per_process)` to make sure every process gets at least a single thread.

fmassa commented:
About the number of processes, one more datapoint from https://github.com/pytorch/vision/pull/818#issuecomment-508855176

> I get the best performance with OMP_NUM_THREADS=1. For --batch-size=32 the epoch time is 6:35.
>
> With OMP_NUM_THREADS=6 it is 1.5x slower.
>
> All batch sizes benefit from OMP_NUM_THREADS=1, but small batch sizes benefit most.

fmassa commented on torch/distributed/launch.py:
```diff
@@ -190,6 +191,13 @@ def parse_args():
 
+def get_omp_num_threads(local_rank, nproc_per_node):
+    avg_threads_per_process = (int)(multiprocessing.cpu_count() / nproc_per_node)
+    residue_threads = multiprocessing.cpu_count() % nproc_per_node
```
question: what's the difference with using `torch.get_num_threads()`?

zhaojuanmao replied:
torch is not imported yet in the launch script, I guess torch will set number threads to be omp_num_threads in application, and torch.get_num_threads() can return that

fmassa commented:
Some more useful debugging information from @andravin on a particular model https://github.com/pytorch/vision/pull/818#issuecomment-510268988

zhaojuanmao commented:
> Some more useful debugging information from @andravin on a particular model [pytorch/vision#818 (comment)](https://github.com/pytorch/vision/pull/818#issuecomment-510268988)

@fmassa thanks for providing one more example for tuning this variable, as you can see this variable should be tuned for some applications, so we print out message to call users' attention to overwrite this variable if needed. 

for this diff, we are recommending better default value based on most of cases, instead of setting default to be number of cpu threads, setting it default to be number of cpu threads/number of processes, expecting new default will bring better performance than old default for most cases, and actually new default makes more senses.

for auto optimal thread tuning possibly need more thoughts and beyond scope of this diff

fmassa commented:
My question is mostly if we should set the default to be 1 instead of num cpus / num gpus.

That's why I was mentioning those threads. Might be worth quickly checking on some standard benchmarks if 1 cpu is better or not

zhaojuanmao commented:
@fmassa I see, I think it makes sense to set omp_num_threads = 1 if your application will mostly run heavy jobs on GPUs, not benefits much from multi CPU threads. I can add this message to script, if that makes sense

zhaojuanmao commented:
@fmassa and @pietern updated diff and message, please let me know whether it is good to go:). thanks

pietern commented:
@zhaojuanmao I think @fmassa's point was to be more conservative by default. Then we would set OMP_NUM_THREADS=1 by default and log the message so users are invited to *increase* it for more performance (because there are idle cores) instead of requiring them to *decrease* it for more performance (because the system is overloaded). What do you think?

zhaojuanmao commented:
> @zhaojuanmao I think @fmassa's point was to be more conservative by default. Then we would set OMP_NUM_THREADS=1 by default and log the message so users are invited to _increase_ it for more performance (because there are idle cores) instead of requiring them to _decrease_ it for more performance (because the system is overloaded). What do you think?

@pietern I'm not sure. Because OMP_NUM_THREADS default is to be number of CPU threads for single process. For multi process, we set it default to be number of CPU threads / number of processes. Both of them align on the same idea => in default make full use of CPU threads -- it is assuming most of use cases, users want that..?

OMP_NUM_THREADS=1 is quite application specific? if we would prefer to set OMP_NUM_THREADS=1 conservatively, maybe we should set OMP_NUM_THREADS=1 in default for single process as well? 

how do you think?

pietern commented:
@zhaojuanmao The default of setting it to the number of CPU cores in the system is what caused performance regressions in the first place. If we truly only were to launch N processes, then setting it to `#cores / N` would be good. But in practice every one of these processes also spawn a few data loader processes, so you can end up with a multiple of that worth of threads on your system. Then, it may be better to choose conservatively and let the user experiment, instead of assuming things about the user's environment / system.

zhaojuanmao commented:
the checks failures are not related to this diff

pietern reviewed (approved):
Good to go.

Can you double check the wording in the error message. The comma after "overloaded" seems unintentional.

pietern reviewed (approved):
Good to go.

Can you double check the wording in the error message. The comma after "overloaded" seems unintentional.

pietern reviewed (approved):
Good to go.

Can you double check the wording in the error message. The comma after "overloaded" seems unintentional.

zhaojuanmao committed:
make OMP_NUM_THREADS default in launch.py (#22501)

Summary:
per https://github.com/pytorch/pytorch/issues/22260, default number of open mp threads are spawned to be the same of number of cores available, for multi processing data parallel cases, too many threads may be spawned and could overload the CPU, resulting in performance regression.

so set OMP_NUM_THREADS = number of CPU processors/number of processes in default to neither overload or waste CPU threads
Pull Request resolved: https://github.com/pytorch/pytorch/pull/22501

Test Plan:
1. without and with this change, example codes result in same result
      python ~/local/fbsource-fbcode/fbcode/caffe2/torch/distributed/launch.py --nproc_per_node=2 pytorch/examples/yanlizhao/distributed_launch_example.py

  Setting OMP_NUM_THREADS environment variable for each process to be: 24, which
  is max(1, num_cpus / num_processes), you can further tune the variable for optimal performance in your application if needed.
  final loss =  tensor(0.5211, device='cuda:0', grad_fn=<MseLossBackward>)

Differential Revision: D16092225

fbshipit-source-id: e03b8d360b0968ecce4b8dbacd6177efda67fdf3
